File: app/src/main/java/com/prey/exceptions/PreyFirebaseCrashlytics.java
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2023 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.exceptions;

import android.content.Context;

import androidx.annotation.NonNull;

import com.google.firebase.FirebaseApp;
import com.google.firebase.crashlytics.FirebaseCrashlytics;

import com.prey.PreyConfig;

public class PreyFirebaseCrashlytics {

    private static PreyFirebaseCrashlytics instance = null;
    private FirebaseCrashlytics crashlytics = null;

    private PreyFirebaseCrashlytics(Context ctx) {
        //initialize FirebaseCrashlytics
        FirebaseApp.initializeApp(ctx);
        FirebaseCrashlytics crashlytics = FirebaseCrashlytics.getInstance();
        crashlytics.setCustomKey("devicekey", PreyConfig.getPreyConfig(ctx).getDeviceId());
        crashlytics.setCustomKey("apikey", PreyConfig.getPreyConfig(ctx).getApiKey());
    }

    public static PreyFirebaseCrashlytics getInstance(Context ctx) {
        if (instance == null)
            instance = new PreyFirebaseCrashlytics(ctx);
        return instance;
    }

    /**
     * Method to report an exception
     */
    public void recordException(@NonNull Throwable throwable) {
        crashlytics.recordException(throwable);
    }
}

File: app/src/main/java/com/prey/exceptions/PreyFirebaseCrashlytics.kt
package com.prey.exceptions

import android.content.Context
import com.google.firebase.FirebaseApp
import com.google.firebase.crashlytics.FirebaseCrashlytics

import com.prey.PreyConfig

/**
 * A private constructor class for handling Firebase Crashlytics.
 * This class is responsible for initializing Firebase Crashlytics and reporting exceptions.
 */
class PreyFirebaseCrashlytics private constructor(context: Context) {

    // Initialize Firebase Crashlytics instance
    private val crashlytics: FirebaseCrashlytics = FirebaseCrashlytics.getInstance()

    /**
     * Initializes Firebase Crashlytics with the provided context.
     * Sets custom keys for device ID and API key.
     */
    init {
        FirebaseApp.initializeApp(context)
        // Set custom keys for device ID and API key
        crashlytics.setCustomKey("devicekey", PreyConfig.getInstance(context).getDeviceId()!!)
        crashlytics.setCustomKey("apikey", PreyConfig.getInstance(context).getApiKey()!!)
    }

    /**
     * Reports an exception to Firebase Crashlytics.
     *
     * @param throwable The exception to be reported.
     */
    fun recordException(throwable: Throwable) {
        crashlytics.recordException(throwable)
    }

    companion object {
        private var instance: PreyFirebaseCrashlytics? = null
        fun getInstance(context: Context): PreyFirebaseCrashlytics {
            return instance ?: synchronized(this) {
                instance ?: PreyFirebaseCrashlytics(context).also { instance = it }
            }
        }
    }
}